subprojects/scala/src/integTest/groovy/org/gradle/scala/compile/CachedScalaCompileIntegrationTest.groovy

```diff
@@ -193,6 +195,52 @@ class CachedScalaCompileIntegrationTest extends AbstractCachedCompileIntegration
         classes.analysisFile.assertIsFile()
     }
 
+    @Requires(TestPrecondition.FIX_TO_WORK_ON_JAVA9) // Zinc cannot do incremental compilation on Java 9, yet
+    def "stale outputs are cleanup up for the first compilation after loading from cache"() {
+        def source1 = file('src/main/scala/Class1.java')
+        source1 << 'class Class1 {}'
+        def source2 = file('src/main/scala/Class2.java')
+        source2 << 'class Class2 {}'
+        def source3 = file('src/main/scala/proto/Class3.java')
+        source3 << 'package proto; class Class3 {}'
+        def class2 = file('build/classes/scala/main/Class2.class')
+        def class1 = file('build/classes/scala/main/Class1.class')
+        def class3 = file('build/classes/scala/main/proto/Class3.class')
+
+        when:
+        withBuildCache().run(compilationTask)
+        then:
+        class1.isFile()
+        class2.isFile()
+        class3.isFile()
+        file(compiledFile).isFile()
+
+        when:
+        run("clean")
+        withBuildCache().run(compilationTask)
+        then:
+        skipped(compilationTask)
+
+        when:
+        assert source3.delete()
+        run(compilationTask)
+        then:
+        executedAndNotSkipped(compilationTask)
+        class1.exists()
+        class2.exists()
+        !class3.exists()
+        !class3.parentFile.exists()
+
+        when:
+        assert source2.delete()
+        run(compilationTask)
```

Comment:
The cache is not involved in this run (the previous run did execute), shouldn't we have a load here, too?

Reply:
I wanted to test if Zinc itself takes care of stale file cleanup when there is no interaction from the cache. There would't be a cache hit since I removed another file.

Should I also add a test for the case when the analysis file is deleted by loading from the cache?

Reply:
I see. I think a separate test would be much better, especially that in the test's name you could describe exactly what you wrote here.

Reply:
I created a separate test case for the stale output removal by Zinc incremental compilation after loading from cache.

This test now also checks if stale class files are removed when loading in a non-clean state from cache.